### ISSUE TITLE
Frequent "ValueError: The parameter mu has invalid values"

### ISSUE THREAD

**Issue by oscarclivio** [YosefLab/scvi-tools#727]
Dear scVI team,

Since the introduction of your custom ZINB and NB distribution classes, it seems that training often feeds NaN values to the inferred NB mean, especially for small datasets. That makes hyperparameter optimization very difficult on these. Here is an example tested with scVI 0.6.4 and 0.6.7 on Ubuntu 18.04. Is there any explanation for this ?

Many thanks.

```
import numpy as np
import torch

torch.manual_seed(0)
np.random.seed(0)

import scvi
from scvi.dataset import CortexDataset
from scvi.inference import UnsupervisedTrainer
from scvi.models.vae import VAE

gene_dataset = CortexDataset()

#np.random.seed(int(time.time()))
#torch.manual_seed(int(time.time()))

model_kwargs = {
    'dropout_rate': 0.519977473649541,
    'n_hidden': 128,
    'n_latent': 6,
    'n_layers': 4,
    'reconstruction_loss': 'zinb'
}
trainer_kwargs = {
    'early_stopping_kwargs': {
        'early_stopping_metric': 'elbo',
        'save_best_state_metric': 'elbo',
        'patience': 30, 'threshold': 0,
        'reduce_lr_on_plateau': True,
        'lr_patience': 15,
        'lr_factor': 0.5
    },
    'n_epochs_kl_warmup': None,
    'use_cuda': True,
    'show_progbar': False,
    'frequency': 1
}
train_func_kwargs = {
    'lr': 0.01,
    'n_epochs': 150
}

model = VAE(gene_dataset.nb_genes, **model_kwargs)
trainer = UnsupervisedTrainer(model=model, gene_dataset=gene_dataset, **trainer_kwargs)
trainer.train(**train_func_kwargs)
```

**Comment by adamgayoso** [comment]
Hi Oscar,

I suspect this has been an issue before we made the distribution classes because we preserved the log likelihood computation. I think this was always a problem and PyTorch has safeguards against an exploding mean parameter during inference. 

I think we have two options.

1. Change that ValueError to a warning
2. Clamp the library size
2a. Check if the library size prior params are suspicious.

**Comment by massonix** [comment]
Dear @adamgayoso,

I am having the same error as @oscarclivio but when running totalVI to denoise protein expression values.

After converting my Seurat object to .h5ad via SeuratDisk, I run the following code:

```{python}
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scanpy as sc
import scvi


path_to_h5ad = "path/to/RNA/expression"
path_to_h5ad_ADT = "path/to/protein/expression"
adata = sc.read_h5ad(path_to_h5ad)
adata_ADT = sc.read_h5ad(path_to_h5ad_ADT)
adata
adata_ADT

adt_expression = adata_ADT.X.toarray()
adt_expression = pd.DataFrame(data = adt_expression, index = adata_ADT.obs_names, columns = adata_ADT.var_names)
adata.obsm["protein_expression"] = adt_expression

adata.layers["counts"] = adata.X.copy()
sc.pp.normalize_total(adata, target_sum = 1e4)
sc.pp.log1p(adata)
adata.raw = adata

sc.pp.highly_variable_genes(adata, n_top_genes = 4000, batch_key = "gem_id")
adata.layers["counts"] = adata.layers["counts"].toarray()

path_to_save_adata = "path/to/save/data"
adata.write_h5ad(filename = path_to_save_adata)
```

Then, I try to run totalVI as follows (after importing packages):

```
scvi.data.setup_anndata(
    adata,
    layer="counts",
    batch_key="gem_id",
    protein_expression_obsm_key="protein_expression"
)
vae = scvi.model.TOTALVI(adata, use_cuda=True, latent_distribution="normal")
vae.train()
```

And at this point I get the error "ValueError: The parameter mu has invalid values".

Could you please let me know which is the best way to send you some test data to reproduce the error?

Thanks a lot for your precious time and help.

Best

Ramon

**Comment by massonix** [comment]
For a bit more of insight, if I find the highest variable genes with `flavor="seurat_v3"`, then I get the following error: 

```
ValueError                                Traceback (most recent call last)
<ipython-input-26-8e24ce27c05e> in <module>
      5     batch_key="gem_id",
      6     subset=True,
----> 7     layer="counts"
      8 )

~/opt/anaconda3/envs/scVI-env/lib/python3.7/site-packages/scanpy/preprocessing/_highly_variable_genes.py in highly_variable_genes(adata, layer, n_top_genes, min_disp, max_disp, min_mean, max_mean, span, n_bins, flavor, subset, inplace, batch_key)
    420             span=span,
    421             subset=subset,
--> 422             inplace=inplace,
    423         )
    424 

~/opt/anaconda3/envs/scVI-env/lib/python3.7/site-packages/scanpy/preprocessing/_highly_variable_genes.py in _highly_variable_genes_seurat_v3(adata, layer, n_top_genes, batch_key, span, subset, inplace)
     82         x = np.log10(mean[not_const])
     83         model = loess(x, y, span=span, degree=2)
---> 84         model.fit()
     85         estimat_var[not_const] = model.outputs.fitted_values
     86         reg_std = np.sqrt(10 ** estimat_var)

_loess.pyx in _loess.loess.fit()

ValueError: b'There are other near singularities as well. 0.090619\n'
```

**Comment by adamgayoso** [comment]
Hi @massonix sorry for the delay. I would first try to reduce the learning rate. Though we have a beta of version 0.8.0 (`pip install scvi-tools==0.8.0b0) which handles the library size a bit differently, mitigating this error.

The HVG error is likely that there are either very lowly expressed genes or very highly expressed genes in your dataset.

**Comment by adamgayoso** [comment]
I'm closing this as the problem should have been fixed in 0.8.X, related to library size computation.
